src/compiler/checker.ts

```diff
@@ -12941,7 +12939,15 @@ namespace ts {
                         continue;
                     }
                 }
-                result.push(getSignatureFromDeclaration(decl));
+                // If this is a function or method declaration, get the accurate minArgumentCount from the @type tag, if present.
+                // If this is a variable or property declaration, apply the @type tag to it
+                // (getTypeForVariableLikeDeclaration()), not to the initializer.

```

Comment:
@sandersn I agree, as written this whole comment isn't that valuable ... I guess what I was aiming to answer is:
- Q: Why are we doing this for function and method declarations?
  A: For optional parameters.
- Q: Why are we excluding contextually-typed kinds?
  A: Because they're already handled elsewhere, and not excluding them would suppress compatibility checks.

I've added a commit with an alternative rewording (based on your suggestions), but I'm very open to going with your preference!